mat: get diagonal of a matrix

as per discussion with @btracey on slack, it'd very handy to be able to extract the diagonal of a matrix for e. g. extracting the variance from a varcov matrix (cf. `numpy.diag` when operating on a matrix)

possibly as a view directly on the matrix:

```go
func (m *Dense) DiagView() SymBandDense {}
```

or

```
func (m *Dense) DiagView() Dense {}
```

or

```go
func (s *SymBandDense) GetDiag(a Matrix)
```

Has this been implemented yet? I've began implementing the above mentioned functionality including updating vector_test.go to provide coverage. At the risk of exposing my naiveté, when I execute the command `go test -run TestVecDenseDiagonalOf`, the following error message is printed:

`inner.go:9:2: use of internal package not allowed`

So, two things, can anyone tell me what I'm doing wrong running the newly implemented tests? and does anyone still care enough about this functionality for me to create a PR?

>when I execute the command `go test -run TestVecDenseDiagonalOf`, the following error message is printed:

>`inner.go:9:2: use of internal package not allowed`

>So, two things, can anyone tell me what I'm doing wrong running the newly implemented tests?

The issue here is likely that you have cloned a github fork of gonum/gonum into a location that is not `$GOPATH/src/gonum.org/v1/gonum`. This will not work, since it will import the code at the locations specified in the import statements. You should use an approach that is equivalent to [this](http://blog.campoy.cat/2014/03/github-and-go-forking-pull-requests-and.html).

I'm working on a PR now.

--- a/mat/diagonal.go
+++ b/mat/diagonal.go
@@ -121,62 +121,63 @@ func (d *DiagDense) Reset() {
 	d.mat.Data = d.mat.Data[:0]
 }
 
-// DiagOf extracts the diagonal of m into the receiver. If m can return
-// its raw representation d will be a view of the diagonal of m. Otherwise
-// the receiver will be a copy of the input's diagonal.
-// The receiver must be min(r, c) long or zero. Otherwise DiagOf will panic.
-func (d *DiagDense) DiagOf(m Matrix) {
-	// reuseAs is not used here since
-	// sensible cases should not allocate,
-	// but would uselessly if we did.
+// DiagFrom copies the diagonal of m into the receiver. The receiver must
+// be min(r, c) long or zero. Otherwise DiagOf will panic.
+func (d *DiagDense) DiagFrom(m Matrix) {
 	n := min(m.Dims())
-	if !d.IsZero() && n != d.n {
-		panic(ErrShape)
-	}
-	d.n = n
+	d.reuseAs(n)
 
+	var vec blas64.Vector
 	switch r := m.(type) {
 	case *DiagDense:
-		d.mat = r.mat
+		vec = r.mat
 	case RawBander:
 		mat := r.RawBand()
-		d.mat = blas64.Vector{
+		vec = blas64.Vector{
 			Inc:  mat.Stride,
 			Data: mat.Data[mat.KL : (n-1)*mat.Stride+mat.KL+1],
 		}
 	case RawMatrixer:
 		mat := r.RawMatrix()
-		d.mat = blas64.Vector{
+		vec = blas64.Vector{
 			Inc:  mat.Stride + 1,
 			Data: mat.Data[:(n-1)*mat.Stride+n],
 		}
 	case RawSymBander:
 		mat := r.RawSymBand()
-		d.mat = blas64.Vector{
+		vec = blas64.Vector{
 			Inc:  mat.Stride,
 			Data: mat.Data[:(n-1)*mat.Stride+1],
 		}
 	case RawSymmetricer:
 		mat := r.RawSymmetric()
-		d.mat = blas64.Vector{
+		vec = blas64.Vector{
 			Inc:  mat.Stride + 1,
 			Data: mat.Data[:(n-1)*mat.Stride+n],
 		}
+	// TODO(kortschak): Add banded triangular handling when the type exists.
 	case RawTriangular:
 		mat := r.RawTriangular()
-		d.mat = blas64.Vector{
+		if mat.Diag == blas.Unit {
+			for i := 0; i < n; i += d.mat.Inc {
+				d.mat.Data[i] = 1
+			}
+			return
+		}
+		vec = blas64.Vector{
 			Inc:  mat.Stride + 1,
 			Data: mat.Data[:(n-1)*mat.Stride+n],
 		}
 	case RawVectorer:
-		d.mat = r.RawVector()
-		d.mat.Data = d.mat.Data[:1]
+		d.mat.Data[0] = r.RawVector().Data[0]
+		return
 	default:
-		d.reuseAs(n)
 		for i := 0; i < n; i++ {
 			d.setDiag(i, m.At(i, i))
 		}
+		return
 	}
+	blas64.Copy(n, vec, d.mat)
 }
 
 // RawBand returns the underlying data used by the receiver represented

--- a/mat/diagonal.go
+++ b/mat/diagonal.go
@@ -40,7 +40,8 @@ type MutableDiagonal interface {
 
 // DiagDense represents a diagonal matrix in dense storage format.
 type DiagDense struct {
-	data []float64
+	mat blas64.Vector
+	n   int
 }
 
 // NewDiagonal creates a new Diagonal matrix with n rows and n columns.
@@ -60,18 +61,19 @@ func NewDiagonal(n int, data []float64) *DiagDense {
 		panic(ErrShape)
 	}
 	return &DiagDense{
-		data: data,
+		mat: blas64.Vector{Data: data, Inc: 1},
+		n:   n,
 	}
 }
 
 // Diag returns the dimension of the receiver.
 func (d *DiagDense) Diag() int {
-	return len(d.data)
+	return d.n
 }
 
 // Dims returns the dimensions of the matrix.
 func (d *DiagDense) Dims() (r, c int) {
-	return len(d.data), len(d.data)
+	return d.n, d.n
 }
 
 // T returns the transpose of the matrix.
@@ -85,41 +87,53 @@ func (d *DiagDense) TTri() Triangular {
 	return TransposeTri{d}
 }
 
+// TBand performs an implicit transpose by returning the receiver inside a
+// TransposeBand.
 func (d *DiagDense) TBand() Banded {
 	return TransposeBand{d}
 }
 
+// Bandwidth returns the upper and lower bandwidths of the matrix.
+// These values are always zero for diagonal matrices.
 func (d *DiagDense) Bandwidth() (kl, ku int) {
 	return 0, 0
 }
 
 // Symmetric implements the Symmetric interface.
 func (d *DiagDense) Symmetric() int {
-	return len(d.data)
+	return d.n
 }
 
 // Triangle implements the Triangular interface.
 func (d *DiagDense) Triangle() (int, TriKind) {
-	return len(d.data), Upper
+	return d.n, Upper
 }
 
+// RawBand returns the underlying data used by the receiver represented
+// as a blas64.Band.
+// Changes to elements in the receiver following the call will be reflected
+// in returned blas64.Band.
 func (d *DiagDense) RawBand() blas64.Band {
 	return blas64.Band{
-		Rows:   len(d.data),
-		Cols:   len(d.data),
+		Rows:   d.n,
+		Cols:   d.n,
 		KL:     0,
 		KU:     0,
-		Stride: 1,
-		Data:   d.data,
+		Stride: d.mat.Inc,
+		Data:   d.mat.Data,
 	}
 }
 
+// RawSymBand returns the underlying data used by the receiver represented
+// as a blas64.SymmetricBand.
+// Changes to elements in the receiver following the call will be reflected
+// in returned blas64.Band.
 func (d *DiagDense) RawSymBand() blas64.SymmetricBand {
 	return blas64.SymmetricBand{
-		N:      len(d.data),
+		N:      d.n,
 		K:      0,
-		Stride: 1,
+		Stride: d.mat.Inc,
 		Uplo:   blas.Upper,
-		Data:   d.data,
+		Data:   d.mat.Data,
 	}
 }

--- a/mat/diagonal.go
+++ b/mat/diagonal.go
@@ -109,6 +109,76 @@ func (d *DiagDense) Triangle() (int, TriKind) {
 	return d.n, Upper
 }
 
+// Reset zeros the length of the matrix so that it can be reused as the
+// receiver of a dimensionally restricted operation.
+//
+// See the Reseter interface for more information.
+func (d *DiagDense) Reset() {
+	// No change of Inc or n to 0 may be
+	// made unless both are set to 0.
+	d.mat.Inc = 0
+	d.n = 0
+	d.mat.Data = d.mat.Data[:0]
+}
+
+// DiagOf extracts the diagonal of m into the receiver. If m can return
+// its raw representation d will be a view of the diagonal of m. Otherwise
+// the receiver will be a copy of the input's diagonal.
+// The receiver must be min(r, c) long or zero. Otherwise DiagOf will panic.
+func (d *DiagDense) DiagOf(m Matrix) {
+	// reuseAs is not used here since
+	// sensible cases should not allocate,
+	// but would uselessly if we did.
+	n := min(m.Dims())
+	if !d.IsZero() && n != d.n {
+		panic(ErrShape)
+	}
+	d.n = n
+
+	switch r := m.(type) {
+	case *DiagDense:
+		d.mat = r.mat
+	case RawBander:
+		mat := r.RawBand()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride,
+			Data: mat.Data[mat.KL : (n-1)*mat.Stride+mat.KL+1],
+		}
+	case RawMatrixer:
+		mat := r.RawMatrix()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride + 1,
+			Data: mat.Data[:(n-1)*mat.Stride+n],
+		}
+	case RawSymBander:
+		mat := r.RawSymBand()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride,
+			Data: mat.Data[:(n-1)*mat.Stride+1],
+		}
+	case RawSymmetricer:
+		mat := r.RawSymmetric()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride + 1,
+			Data: mat.Data[:(n-1)*mat.Stride+n],
+		}
+	case RawTriangular:
+		mat := r.RawTriangular()
+		d.mat = blas64.Vector{
+			Inc:  mat.Stride + 1,
+			Data: mat.Data[:(n-1)*mat.Stride+n],
+		}
+	case RawVectorer:
+		d.mat = r.RawVector()
+		d.mat.Data = d.mat.Data[:1]
+	default:
+		d.reuseAs(n)
+		for i := 0; i < n; i++ {
+			d.setDiag(i, m.At(i, i))
+		}
+	}
+}
+
 // RawBand returns the underlying data used by the receiver represented
 // as a blas64.Band.
 // Changes to elements in the receiver following the call will be reflected
@@ -136,4 +206,31 @@ func (d *DiagDense) RawSymBand() blas64.SymmetricBand {
 		Uplo:   blas.Upper,
 		Data:   d.mat.Data,
 	}
+}
+
+// reuseAs resizes an empty diagonal to a r×r diagonal,
+// or checks that a non-empty matrix is r×r.
+func (d *DiagDense) reuseAs(r int) {
+	if r == 0 {
+		panic(ErrZeroLength)
+	}
+	if d.IsZero() {
+		d.mat = blas64.Vector{
+			Inc:  1,
+			Data: use(d.mat.Data, r),
+		}
+		d.n = r
+		return
+	}
+	if r != d.n {
+		panic(ErrShape)
+	}
+}
+
+// IsZero returns whether the receiver is zero-sized. Zero-sized vectors can be the
+// receiver for size-restricted operations. DiagDenses can be zeroed using Reset.
+func (d *DiagDense) IsZero() bool {
+	// It must be the case that d.Dims() returns
+	// zeros in this case. See comment in Reset().
+	return d.mat.Inc == 0
 }

--- a/mat/diagonal_test.go
+++ b/mat/diagonal_test.go
@@ -7,6 +7,8 @@ package mat
 import (
 	"reflect"
 	"testing"
+
+	"gonum.org/v1/gonum/blas/blas64"
 )
 
 func TestNewDiagonal(t *testing.T) {
@@ -20,7 +22,8 @@ func TestNewDiagonal(t *testing.T) {
 			data: []float64{1, 2, 3, 4, 5, 6},
 			n:    6,
 			mat: &DiagDense{
-				data: []float64{1, 2, 3, 4, 5, 6},
+				mat: blas64.Vector{Inc: 1, Data: []float64{1, 2, 3, 4, 5, 6}},
+				n:   6,
 			},
 			dense: NewDense(6, 6, []float64{
 				1, 0, 0, 0, 0, 0,
@@ -49,6 +52,75 @@ func TestNewDiagonal(t *testing.T) {
 		}
 		if !Equal(band, test.dense) {
 			t.Errorf("unexpected value via mat.Equal(band, dense) for test %d:\ngot:\n% v\nwant:\n% v", i, Formatted(band), Formatted(test.dense))
+		}
+	}
+}
+
+func TestDiagonalStride(t *testing.T) {
+	for _, test := range []struct {
+		diag  *DiagDense
+		dense *Dense
+	}{
+		{
+			diag: &DiagDense{
+				mat: blas64.Vector{Inc: 1, Data: []float64{1, 2, 3, 4, 5, 6}},
+				n:   6,
+			},
+			dense: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			diag: &DiagDense{
+				mat: blas64.Vector{Inc: 2, Data: []float64{
+					1, 0,
+					2, 0,
+					3, 0,
+					4, 0,
+					5, 0,
+					6,
+				}},
+				n: 6,
+			},
+			dense: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			diag: &DiagDense{
+				mat: blas64.Vector{Inc: 5, Data: []float64{
+					1, 0, 0, 0, 0,
+					2, 0, 0, 0, 0,
+					3, 0, 0, 0, 0,
+					4, 0, 0, 0, 0,
+					5, 0, 0, 0, 0,
+					6,
+				}},
+				n: 6,
+			},
+			dense: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+	} {
+		if !Equal(test.diag, test.dense) {
+			t.Errorf("unexpected value via mat.Equal for stride %d: got: %v want: %v",
+				test.diag.mat.Inc, test.diag, test.dense)
 		}
 	}
 }

--- a/mat/diagonal_test.go
+++ b/mat/diagonal_test.go
@@ -5,6 +5,7 @@
 package mat
 
 import (
+	"math"
 	"reflect"
 	"testing"
 
@@ -121,6 +122,202 @@ func TestDiagonalStride(t *testing.T) {
 		if !Equal(test.diag, test.dense) {
 			t.Errorf("unexpected value via mat.Equal for stride %d: got: %v want: %v",
 				test.diag.mat.Inc, test.diag, test.dense)
+		}
+	}
+}
+
+func TestDiagOf(t *testing.T) {
+	for i, test := range []struct {
+		mat  Matrix
+		want *Dense
+	}{
+		{
+			mat: NewDiagonal(6, []float64{1, 2, 3, 4, 5, 6}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat: NewBandDense(6, 6, 1, 1, []float64{
+				math.NaN(), 1, math.NaN(),
+				math.NaN(), 2, math.NaN(),
+				math.NaN(), 3, math.NaN(),
+				math.NaN(), 4, math.NaN(),
+				math.NaN(), 5, math.NaN(),
+				math.NaN(), 6, math.NaN(),
+			}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat: NewDense(6, 6, []float64{
+				1, math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), 2, math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), 3, math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), 4, math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), 5, math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 6,
+			}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat: NewDense(6, 4, []float64{
+				1, math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), 2, math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), 3, math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), 4,
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+			}),
+			want: NewDense(4, 4, []float64{
+				1, 0, 0, 0,
+				0, 2, 0, 0,
+				0, 0, 3, 0,
+				0, 0, 0, 4,
+			}),
+		},
+		{
+			mat: NewDense(4, 6, []float64{
+				1, math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), 2, math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), 3, math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), 4, math.NaN(), math.NaN(),
+			}),
+			want: NewDense(4, 4, []float64{
+				1, 0, 0, 0,
+				0, 2, 0, 0,
+				0, 0, 3, 0,
+				0, 0, 0, 4,
+			}),
+		},
+		{
+			mat: NewSymBandDense(6, 1, []float64{
+				1, math.NaN(),
+				2, math.NaN(),
+				3, math.NaN(),
+				4, math.NaN(),
+				5, math.NaN(),
+				6, math.NaN(),
+			}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat: NewSymDense(6, []float64{
+				1, math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), 2, math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), 3, math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), 4, math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), 5, math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 6,
+			}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat: NewTriDense(6, Upper, []float64{
+				1, math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), 2, math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), 3, math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), 4, math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), 5, math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 6,
+			}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat: NewTriDense(6, Lower, []float64{
+				1, math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), 2, math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), 3, math.NaN(), math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), 4, math.NaN(), math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), 5, math.NaN(),
+				math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 6,
+			}),
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+		{
+			mat:  NewVecDense(6, []float64{1, 2, 3, 4, 5, 6}),
+			want: NewDense(1, 1, []float64{1}),
+		},
+		{
+			mat: &basicMatrix{
+				mat: blas64.General{
+					Rows:   6,
+					Cols:   6,
+					Stride: 6,
+					Data: []float64{
+						1, math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+						math.NaN(), 2, math.NaN(), math.NaN(), math.NaN(), math.NaN(),
+						math.NaN(), math.NaN(), 3, math.NaN(), math.NaN(), math.NaN(),
+						math.NaN(), math.NaN(), math.NaN(), 4, math.NaN(), math.NaN(),
+						math.NaN(), math.NaN(), math.NaN(), math.NaN(), 5, math.NaN(),
+						math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 6,
+					},
+				},
+				capRows: 6,
+				capCols: 6,
+			},
+			want: NewDense(6, 6, []float64{
+				1, 0, 0, 0, 0, 0,
+				0, 2, 0, 0, 0, 0,
+				0, 0, 3, 0, 0, 0,
+				0, 0, 0, 4, 0, 0,
+				0, 0, 0, 0, 5, 0,
+				0, 0, 0, 0, 0, 6,
+			}),
+		},
+	} {
+		var got DiagDense
+		got.DiagOf(test.mat)
+		if !Equal(&got, test.want) {
+			r, c := test.mat.Dims()
+			t.Errorf("unexpected value via mat.Equal for %d×%d %T test %d:\ngot:\n% v\nwant:\n% v",
+				r, c, test.mat, i, Formatted(&got), Formatted(test.want))
 		}
 	}
 }

--- a/mat/diagonal_test.go
+++ b/mat/diagonal_test.go
@@ -126,7 +126,7 @@ func TestDiagonalStride(t *testing.T) {
 	}
 }
 
-func TestDiagOf(t *testing.T) {
+func TestDiagFrom(t *testing.T) {
 	for i, test := range []struct {
 		mat  Matrix
 		want *Dense
@@ -313,7 +313,7 @@ func TestDiagOf(t *testing.T) {
 		},
 	} {
 		var got DiagDense
-		got.DiagOf(test.mat)
+		got.DiagFrom(test.mat)
 		if !Equal(&got, test.want) {
 			r, c := test.mat.Dims()
 			t.Errorf("unexpected value via mat.Equal for %d×%d %T test %d:\ngot:\n% v\nwant:\n% v",

--- a/mat/index_bound_checks.go
+++ b/mat/index_bound_checks.go
@@ -238,16 +238,16 @@ func (d *DiagDense) At(i, j int) float64 {
 }
 
 func (d *DiagDense) at(i, j int) float64 {
-	if uint(i) >= uint(len(d.data)) {
+	if uint(i) >= uint(d.n) {
 		panic(ErrRowAccess)
 	}
-	if uint(j) >= uint(len(d.data)) {
+	if uint(j) >= uint(d.n) {
 		panic(ErrColAccess)
 	}
 	if i != j {
 		return 0
 	}
-	return d.data[i]
+	return d.mat.Data[i*d.mat.Inc]
 }
 
 // SetDiag sets the element at row i, column i to the value v.
@@ -257,8 +257,8 @@ func (d *DiagDense) SetDiag(i int, v float64) {
 }
 
 func (d *DiagDense) setDiag(i int, v float64) {
-	if uint(i) >= uint(len(d.data)) {
+	if uint(i) >= uint(d.n) {
 		panic(ErrRowAccess)
 	}
-	d.data[i] = v
+	d.mat.Data[i*d.mat.Inc] = v
 }

--- a/mat/index_no_bound_checks.go
+++ b/mat/index_no_bound_checks.go
@@ -238,10 +238,10 @@ func (s *SymBandDense) set(i, j int, v float64) {
 
 // At returns the element at row i, column j.
 func (d *DiagDense) At(i, j int) float64 {
-	if uint(i) >= uint(len(d.data)) {
+	if uint(i) >= uint(d.n) {
 		panic(ErrRowAccess)
 	}
-	if uint(j) >= uint(len(d.data)) {
+	if uint(j) >= uint(d.n) {
 		panic(ErrColAccess)
 	}
 	return d.at(i, j)
@@ -251,18 +251,18 @@ func (d *DiagDense) at(i, j int) float64 {
 	if i != j {
 		return 0
 	}
-	return d.data[i]
+	return d.mat.Data[i*d.mat.Inc]
 }
 
 // SetDiag sets the element at row i, column i to the value v.
 // It panics if the location is outside the appropriate region of the matrix.
 func (d *DiagDense) SetDiag(i int, v float64) {
-	if uint(i) >= uint(len(d.data)) {
+	if uint(i) >= uint(d.n) {
 		panic(ErrRowAccess)
 	}
 	d.setDiag(i, v)
 }
 
 func (d *DiagDense) setDiag(i int, v float64) {
-	d.data[i] = v
+	d.mat.Data[i*d.mat.Inc] = v
 }

--- a/mat/list_test.go
+++ b/mat/list_test.go
@@ -482,9 +482,10 @@ func makeCopyOf(a Matrix) Matrix {
 		return m
 	case *DiagDense:
 		d := &DiagDense{
-			data: make([]float64, len(t.data)),
+			mat: blas64.Vector{Inc: 1, Data: make([]float64, len(t.mat.Data))},
+			n:   t.n,
 		}
-		copy(d.data, t.data)
+		copy(d.mat.Data, t.mat.Data)
 		return d
 	}
 }
@@ -618,10 +619,12 @@ var testMatrices = []Matrix{
 	&SymDense{},
 	NewTriDense(3, true, nil),
 	NewTriDense(3, false, nil),
-	&VecDense{mat: blas64.Vector{Inc: 1}},
-	&DiagDense{},
 	&basicVector{},
+	&VecDense{mat: blas64.Vector{Inc: 1}},
 	&VecDense{mat: blas64.Vector{Inc: 10}},
+	&DiagDense{},
+	&DiagDense{mat: blas64.Vector{Inc: 1}},
+	&DiagDense{mat: blas64.Vector{Inc: 10}},
 	&basicMatrix{},
 	&basicSymmetric{},
 	&basicTriangular{cap: 3, mat: blas64.Triangular{N: 3, Stride: 3, Uplo: blas.Upper}},

